bioframe/extras.py

```diff
@@ -499,3 +503,161 @@ def pair_by_distance(
+    merge_runs
+    """
+    ck, sk, ek = _get_default_colnames() if cols is None else cols
+
+    if not allow_overlaps and len(ops.overlap(df, df)) > len(df):
+        raise ValueError("Not a proper bedGraph: found overlapping intervals.")
+
+    result = []
+    where = np.flatnonzero
+    n_runs = 0
+
+    for _, group in df.groupby(ck, sort=False):
+        group = group.sort_values([sk, ek])
+
+        # Find runs of values
+        values = group[col].to_numpy()
+        isnumeric = np.issubdtype(values.dtype, np.number)
+
+        if isnumeric:
+            run_starts = np.r_[
+                0,
+                where(~np.isclose(values[1:], values[:-1], equal_nan=True)) + 1
+            ]
+        else:
+            run_starts = np.r_[0, where(values[1:] != values[:-1]) + 1]
+
+        run_lengths = np.diff(np.r_[run_starts, len(values)])
+        run_ends = run_starts + run_lengths
+
+        # Assign run numbers to intervals
+        if reset_counter:
+            n_runs = 0
+        group[run_col] = pd.NA
+        j = group.columns.get_loc(run_col)
+        for lo, hi in zip(run_starts, run_ends):
+            group.iloc[lo : hi + 1, j] = n_runs
+            n_runs += 1
+
+        result.append(group)
+
+    return pd.concat(result)
+
+
+def merge_runs(
+    df: pd.DataFrame,
+    col: str,
+    *,
+    allow_overlaps: bool = False,
+    agg: Optional[dict] = None,
+    cols: Optional[Tuple[str, str, str]] = None,
+) -> pd.DataFrame:
+    """
+    Merge runs of immediately consecutive intervals sharing the same value of
+    ``col``.
+
+    Parameters
+    ----------
+    df : DataFrame
+        A bioframe dataframe.
+    col : str
+        The column to compress runs of values for.
+    allow_overlaps : bool, optional [default: False]
+        If True, allow intervals in ``df`` to overlap. This may cause
+        unexpected results.
+    agg : dict, optional [default: None]
+        A dictionary of additional column names and aggregation functions to
+        apply to each run. Takes the format:
+            {'agg_name': ('column_name', 'agg_func')}
+
+    Returns
+    -------
+    pandas.DataFrame
+        Dataframe with consecutive intervals in the same run merged.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     'chrom': ['chr1', 'chr1', 'chr1', 'chr1', 'chr1', 'chr1'],
+    ...     'start': [0, 100, 200, 300, 400, 500],
+    ...     'end': [100, 200, 300, 400, 500, 600],
+    ...     'value': [1, 1, 1, 2, 2, 2],
+    ... })
+
+    >>> merge_runs(df, 'value')
+        chrom  start  end  value
+    0   chr1      0  300      1
+    1   chr1    300  600      2
+
+    >>> merge_runs(df, 'value', agg={'sum': ('value', 'sum')})
+        chrom  start  end  value  sum
+    0   chr1      0  300      1    3
+    1   chr1    300  600      2    6
+
+    See Also
+    --------
+    mark_runs
+    """
+    ck, sk, ek = _get_default_colnames() if cols is None else cols
+
+    if agg is None:
+        agg = {}
```

Comment:
This could be set as a default parameter

Reply:
Mutable defaults in Python are an [antipattern](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/)

Reply:
Interesting... wonder if that's by language design :sweat_smile: